st2common/st2common/util/config_loader.py

```diff
@@ -101,16 +99,19 @@ def _get_values_for_config(self, config_schema_db, config_db):
         return config
 
     @staticmethod
-    def _get_object_property_schema(object_schema, init_additional_properties=None):
+    def _get_object_property_schema(object_schema, additional_properties_keys=None):
+        """
+        Create a schema for an object property using both additionalProperties and properties.
+
+        :rtype: ``dict``
+        """
+        property_schema = {}
         additional_properties = object_schema.get("additionalProperties", {})
+        # additionalProperties can be a boolean or a dict
         if additional_properties and isinstance(additional_properties, dict):
-            property_schema = defaultdict(lambda: additional_properties)
-        else:
-            property_schema = {}
-        if init_additional_properties:
-            # ensure that these keys are present in the object (vs just defaultdict)
-            for key in init_additional_properties:
-                property_schema.__missing__(key)
+            # ensure that these keys are present in the object
+            for key in additional_properties_keys:
```

Comment:
Thanks, if that works it should be much more straight forward to understand the code now - well, at least for me :)

Reply:
And re - using copy - I actually verified, if we didn't use copy with ``defaultdict`` version, we also don't need to use it here - it's returning by reference version in both scenarios.

And since I believe we indeed never manipulate those values, only read / access them, copy is probably not needed.

Reply:
Yes. We modify the `config` object, but we don't modify (and probably should not modify) the schema itself. So, references are perfect in this case.